pkg/adapter/common/router/router.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright (c) 2021 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package router
+
+import (
+	"net/http"
+	"sync"
+)
+
+// Router routes incoming HTTP requests to the adequate handler based on their
+// URL path.
+type Router struct {
+	// map of URL path to HTTP handler
+	handlers sync.Map
+}
+
+// Check that Router implements http.Handler.
+var _ http.Handler = (*Router)(nil)
+
+// RegisterPath registers a HTTP handler for serving requests at the given URL path.
+func (r *Router) RegisterPath(urlPath string, h http.Handler) {
+	r.handlers.Store(urlPath, h)
+}
+
+// DeregisterPath de-registers the HTTP handler for the given URL path.
+func (r *Router) DeregisterPath(urlPath string) {
+	r.handlers.Delete(urlPath)
+}
```

Comment:
Wasn't thinking of Knative per-se, but meshes like Istio which do rely on it.

Reply:
Could you please share resources about it?

Reply:
The case that I'm thinking of was: https://istio.io/latest/docs/ops/configuration/mesh/app-health-check/

Reply:
I see that all Knative Services inject this probe implicitly (_this source is deployed as a Knative Service_):
```yaml
        readinessProbe:
          exec:
            command:
            - /ko-app/queue
            - -probe-period
            - "0"
          failureThreshold: 3
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 10
```

The behaviour is explained at https://github.com/knative/serving/blob/master/docs/runtime-contract.md#meta-requests

> If not provided, container startup and **listening on the declared HTTP** socket is considered sufficient to declare the container "ready" and "live".

Do you think we need to do more? Would reply with 200 be better than probing that the socket on the HTTP port is listening? I'm asking because we didn't have a probe before this PR. There was a "/healthz" endpoint but nothing was configured to probe it. I don't mind adding it back.

Reply:
Okay so it is mainly leveraging the current default behavior from Kubernetes. Don't think we'd need to do more unless you want to ensure that the source is able to connect to the SNS endpoint, but suspect this is more useful for targets.

Reply:
Right it might not be very useful here since there is no connection to SNS involved. The controller subscribes the source as an endpoint, and SNS _sends_ HTTP requests to us, so "HTTP is listening" should be a good indicator that we are ready to serve traffic, but I might be wrong and I'm happy to revisit now or at a later point.

Reply:
Yup, and is something to keep in mind given the other sources that have the health check endpoints provided.